src/assets/scripts/client/aircraft/AircraftModel.js

```diff
@@ -1862,32 +1862,102 @@ export default class AircraftModel {
+                        }
+                    }
+
+                    // ...so descend to comply with VNAV restriction
+                    return this._calculateTargetedAltitudeVnavDescent(nextAltitudeMaximumWaypoint);
+                }
+            }
+        }
+        else {
+            // we want to climb...
+            if (!maximumAltitudeExists || this.mcp.altitude <= nextAltitudeMaximumWaypoint.altitudeMaximum) {
+                // ... and there is nothing that can stop us.
+                return this.mcp.altitude;
+            }
+
+            if (this.altitude > nextAltitudeMaximumWaypoint.altitudeMaximum) {
+                // .. but we are too high and have to comply with NAV restriction
                 return this._calculateTargetedAltitudeVnavDescent(nextAltitudeMaximumWaypoint);
             }
 
-            return this._calculateTargetedAltitudeVnavClimb(nextAltitudeMinimumWaypoint);
-        } else if (maximumAltitudeExists) {
-            return this._calculateTargetedAltitudeVnavDescent(nextAltitudeMaximumWaypoint);
-        } else if (minimumAltitudeExists) {
-            return this._calculateTargetedAltitudeVnavClimb(nextAltitudeMinimumWaypoint);
+            if (minimumAltitudeExists) {
+                if (this.mcp.altitude < nextAltitudeMinimumWaypoint.altitudeMinimum) {
+                    // we are too low but we are prioritizing clearance over VNAV restriction
+                    return this.mcp.altitude;
+                }
+
+                if (this.altitude < nextAltitudeMinimumWaypoint.altitudeMinimum) {
+                    // we are too low ...
+                    if (nextAltitudeMaximumWaypoint.altitudeMaximum < nextAltitudeMinimumWaypoint.altitudeMinimum) {
+                        // the maximum altitude is below the minimal altiude, check if we can climb all the way up
+                        // without violating VNAV restrictions.
+                        const firstWaypoint = this._findFirstWaypoint(this.fms.waypoints, nextAltitudeMinimumWaypoint, nextAltitudeMaximumWaypoint);
+
+                        if (firstWaypoint.name === nextAltitudeMaximumWaypoint.name) {
+                            // ... but we can not climb all the way up yet
+                            return nextAltitudeMaximumWaypoint.altitudeMaximum;
+                        }
+                    }
+
+                    // ... climb to comply with VNAV restriction
+                    return this._calculateTargetedAltitudeVnavClimb(nextAltitudeMinimumWaypoint);
+                }
+            }
         }
     }
 
+    /**
+     * Takes two waypoints and returns the waypoint that comes first in the list of waypoints
+     *
+     * @for AircraftModel
+     * @method _findFirstWaypoint
+     * @param waypoints
+     * @param waypointA {WaypointModel}
+     * @param waypointB {WaypointModel}
+     * @return waypointA or waypointB {WaypointModel}
+     */
+    _findFirstWaypoint(waypoints, waypointA, waypointB) {
```

Comment:
may have missed where this is getting called, but I feel like this logic should probably live within the `fms` instead

Reply:
called in `_calculateTargetedAltitudeVnav()`